### DISCUSSION TITLE
Last value queue not created when using ASP.NET Core DI

### DISCUSSION THREAD

**Discussion by pgarbo** [Havret/dotnet-activemq-artemis-client#407] [Q&A] [answered]
Hi,

I have noticed that when queue is created using ASP.NET Core DI with ConfigureTopology extension, even if LastValueKey is set, it is not being set in the actual queue. It works fine with a classic topology manager.
Is there anything wrong with my code?

```
services.AddActiveMq("synergy-broker", new[] { endpoint })
            .ConfigureConnectionFactory((provider, factory) =>
            {
                factory.LoggerFactory = provider.GetService<ILoggerFactory>();
                factory.RecoveryPolicy = RecoveryPolicyFactory.ExponentialBackoff(initialDelay: TimeSpan.FromSeconds(1), maxDelay: TimeSpan.FromSeconds(30), retryCount: 5);
                factory.MessageIdPolicyFactory = MessageIdPolicyFactory.GuidMessageIdPolicy;
                factory.AutomaticRecoveryEnabled = true;

            }).ConfigureConnection((_, connection) =>
            {
                connection.ConnectionClosed += (_, args) =>
                {
                    Console.WriteLine($"Connection closed: ClosedByPeer={args.ClosedByPeer}, Error={args.Error}");
                };
                connection.ConnectionRecovered += (_, args) =>
                {
                    Console.WriteLine($"Connection recovered: Endpoint={args.Endpoint}");
                };
                connection.ConnectionRecoveryError += (_, args) =>
                {
                    Console.WriteLine($"Connection recovered error: Exception={args.Exception}");
                };
            }).ConfigureTopology(async (_, topologyManager) =>
            {
                await topologyManager.DeclareQueueAsync(new QueueConfiguration
                {
                    Address = appListQueueName,
                    Name = appListQueueName,
                    Durable = true,
                    AutoCreateAddress = true,
                    NonDestructive = true,
                    RoutingType = RoutingType.Anycast,
                    LastValueKey = ActiveMqConstants.LastValueKey
                });
                await topologyManager.DeclareQueueAsync(new QueueConfiguration
                {
                    Address = appDeployedQueueName,
                    Name = appDeployedQueueName,
                    Durable = true,
                    AutoCreateAddress = true,
                    NonDestructive = false,
                    RoutingType = RoutingType.Multicast,
                });
            })
            .AddProducer<LatestAppListAnycastProducer>(appListQueueName, RoutingType.Anycast)
            .AddProducer<AppDeployedMulticastProducer>(appDeployedQueueName, RoutingType.Multicast)
            .EnableAddressDeclaration()
            .EnableQueueDeclaration();
```

**Comment by pgarbo** [comment]
I did some investigation and it turns out that ConfigureTopology callback is called too late - the queue is already created probably by AddProducer earlier. Is it a bug or do I need to change my code? I tried with EnableQueueDeclaration(false) but it does not change anything.

**Comment by Havret** [comment, marked as answer]
Hi @pgarbo,

Your code looks fine. But if you're trying to built what you described here https://github.com/Havret/dotnet-activemq-artemis-client/discussions/387 you don't need to declare the second queue - `appDeployedQueueName`. Each of your consumers should subscribe to address and create it's own queue dynamically. Additionally `EnableQueueDeclaration` is redundant here.

I hope that helps,
Havret

**Reply by pgarbo** [reply]
Hi @Havret , 

thanks for the suggestions, but there's still an issue with appListQueueName in ConfigureTopology - despite I set LastValueKey, the queue created has these attributes:
Last value :  false
Last value key: empty
![image](https://user-images.githubusercontent.com/16103019/208069397-60e95213-86ab-4295-88cd-5c1761ba4b1c.png)

You can replicate this easily. Looks like a bug with ConfigureTopology method. when I call DeclareQueueAsync from ConfigureConnection, then it is fine but this is non async callback.

Pawel

**Reply by Havret** [reply]
Hi [pgarbo](https://github.com/pgarbo), 

Yes, you're right. Configure topology is called after the producer is already created. As you configured it as an anycast producer it creates its own queue on attach. Because there is a bug in Artemis (you cannot set LastValueKey on queue that already exists) your queue declaration statement has no effect.

What you could do is to create your producer without specifying the routing type:

```csharp
AddProducer<LatestAppListAnycastProducer>(appListQueueName)
```
This should do the trick:

![image](https://user-images.githubusercontent.com/9103861/208151977-a69bc688-c4c5-4f77-931e-25d0cb597d22.png)

From my side I could create a fix that would change the order in which resources are declared/created. 

Best,
Havret

**Reply by pgarbo** [reply]
Hi, the workaround worked, thanks. The fix would be nice though :) 

thanks,
Pawel

**Reply by Havret** [reply]
Fix released as part of 2.12.0.
